torchsparse/nn/functional/downsample.py

```diff
@@ -1,70 +1,43 @@
+    if all(stride[k] in [1, kernel_size[k]] for k in range(3)):
+        sample_stride = torch.tensor(sample_stride,
+                                     dtype=torch.int,
+                                     device=coords.device).unsqueeze(0)
+        coords = coords.clone()
+        coords[:, :3] = coords[:, :3] // sample_stride * sample_stride
     else:
-        if isinstance(kernel_size, int):
-            # ratio is a permutation of [1, 1, kernel_size]
-            direct_downsample = (kernel_size == ratio.prod().item()) & \
-                (torch.sum(ratio == kernel_size) == 1).item()
-        else:
-            direct_downsample = False
+        offsets = get_kernel_offsets(kernel_size,
+                                     tensor_stride,
+                                     device=coords.device)
 
-    if direct_downsample:
-        _ratio = ratio * tensor_stride
-        new_coords = torch.cat(
-            [coords[:, :3] // _ratio * _ratio, coords[:, 3:]], 1)
-        return torch.unique(new_coords, dim=0)
-    else:
-        kernel_region = KernelRegion(kernel_size, tensor_stride, dilation=1)
-        # kernel volume x 3
-        kernel_offset = kernel_region.get_kernel_offset().to(coords.device)
-        new_coords = coords[:, :3].unsqueeze(1).repeat(
-            1, kernel_offset.size(0), 1) + kernel_offset
-        # (N x kernel volume) x 4
-        new_coords = torch.cat([
-            coords[:, 3:].repeat(1, kernel_offset.size(0)).view(-1, 1),
-            new_coords.view(-1, 3)
-        ],
-                               dim=1)
-        new_ts = tensor_stride * ratio
-        # only keep these coordinates that is multiple of new_ts.
-        if isinstance(new_ts, torch.Tensor):
-            new_ts = new_ts[0]
-            new_coords = new_coords[
-                (new_coords[:, 1] % new_ts[0].item() == 0) & (new_coords[:, 2] % new_ts[1].item() == 0) & \
-                (new_coords[:, 3] % new_ts[2].item() == 0)
-            ]
-        else:
-            new_coords = new_coords[
-                (new_coords[:, 1] % new_ts == 0) & (new_coords[:, 2] % new_ts == 0) & \
-                (new_coords[:, 3] % new_ts == 0)
-            ]
-        new_coords = new_coords[(new_coords[:, 1] >= 0)
-                                & (new_coords[:, 2] >= 0) &
-                                (new_coords[:, 3] >= 0)]
-        # filter out duplicates
-        new_coords = torch.unique(new_coords, dim=0)
-        new_coords = new_coords[:, [1, 2, 3, 0]]
-        return new_coords
+        xyz = coords[:, :3].unsqueeze(1).repeat(1, offsets.size(0), 1) + offsets
+        b = coords[:, 3:].repeat(1, offsets.size(0))
+        coords = torch.cat([xyz.view(-1, 3), b.view(-1, 1)], dim=1)
+
+        mask = ((coords[:, 0] % sample_stride[0] == 0)
+                & (coords[:, 1] % sample_stride[1] == 0)
+                & (coords[:, 2] % sample_stride[2] == 0))
+        coords = coords[mask]
+
+    coords = torch.unique(coords, dim=0)
```

Comment:
That's a very good point! I have reordered `coords` to preserve this property. It would be great to add comments before this sort of design choice to let others understand the thinking process.